packages/workspace/src/browser/workspace-service.ts

```diff
@@ -57,18 +67,149 @@ export class WorkspaceService implements FrontendApplicationContribution {
+            }
+
+            for (const rootBeingWatched of Object.keys(this.rootWatchers)) {
+                if (roots.indexOf(rootBeingWatched) < 0) {
+                    this.stopWatch(rootBeingWatched);
+                }
+            }
+            for (const rootToWatch of roots) {
+                if (!this.rootWatchers[rootToWatch]) {
+                    this.startWatch(await this.toValidRoot(rootToWatch));
+                }
+            }
+
+            if (!this._workspaceDescription.workspaceConfigFile) {
+                this._workspaceDescription.workspaceConfigFile = await this.getWorkspaceConfigFileStat();
+            }
         }
+        this.onWorkspaceChangeEmitter.fire(this._workspaceDescription);
+    }
+
+    protected async setupRoots(): Promise<void> {
+        const rootUri = await this.server.getWorkspace();
+        this._root = await this.toValidRoot(rootUri);
+        this._workspaceDescription.root = this._root;
         this.deferredRoot.resolve(this._root);
+
+        this._workspaceDescription.roots = [];
+        await this.preferences.ready;
+        if (this._root) {
+            if (this.preferences['workspace.supportMultiRootWorkspace']) {
+                const roots = await this.getRoots();
+                if (roots.length === 0) { // roots.json is avaialble however no roots are specified
+                    this.startWatch(this._root);
+                } else {
+                    for (const r of roots) {
+                        this.startWatch(await this.toValidRoot(r));
+                    }
+                }
+            } else {
+                this.startWatch(this._root);
+            }
+        }
+
+        if (this._root && this.preferences['workspace.supportMultiRootWorkspace']) {
+            this._workspaceDescription.workspaceConfigFile = await this.getWorkspaceConfigFileStat();
+        }
     }
 
-    protected updateTitle(uri: URI): void {
-        document.title = uri.displayName;
+    protected async getRoots(): Promise<string[]> {
+        const configUri = await this.getWorkspaceConfigFileUri();
+        if (configUri) {
+            const uriStr = configUri.path.toString();
+            if (await this.fileSystem.exists(uriStr)) {
+                const content = await this.fileSystem.resolveContent(uriStr);
+                if (content) {
+                    const roots = JSON.parse(content.content).roots || [];
+                    return this._root && roots.length === 0 ? [this._root.uri] : roots;
+                }
+            }
+            return [this._root!.uri];
+        }
+        return [];
+    }
+
+    async getWorkspaceConfigFileUri(): Promise<URI | undefined> {
+        if (this._root) {
+            const rootUri = new URI(this._root.uri);
+            return new URI(path.resolve(rootUri.path.toString(), '.theia', 'roots.json')).withScheme('file');
+        }
+    }
+
+    private async getWorkspaceConfigFileStat(): Promise<FileStat | undefined> {
```

Comment:
we should merge this method with `getRoots`, `FileSystem.resolveContent` already returns `FileStat`, no need to fetch it again